ssl/quic/json_enc.c

```diff
@@ -0,0 +1,721 @@
+/*
+ * Copyright 2023 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include "internal/json_enc.h"
+#include "internal/nelem.h"
+#include <string.h>
+#include <math.h>
+
+/*
+ * wbuf
+ * ====
+ */
+static int wbuf_flush(struct json_write_buf *wbuf);
+
+static int wbuf_init(struct json_write_buf *wbuf, BIO *bio, size_t alloc)
+{
+    wbuf->buf = OPENSSL_malloc(alloc);
+    if (wbuf->buf == NULL)
+        return 0;
+
+    wbuf->cur   = 0;
+    wbuf->alloc = alloc;
+    wbuf->bio   = bio;
+    return 1;
+}
+
+static void wbuf_cleanup(struct json_write_buf *wbuf)
+{
+    OPENSSL_free(wbuf->buf);
+    wbuf->buf   = NULL;
+    wbuf->alloc = 0;
+}
+
+static void wbuf_set_bio(struct json_write_buf *wbuf, BIO *bio)
+{
+    wbuf->bio = bio;
```

Comment:
Should this free any pre-existing BIO? Should this be called set0?

Reply:
Renamed

Reply:
What about my first question - should this free a pre-existing BIO?

Reply:
Sorry for answering here: I think the code is OK as-is. The `wbuf->bio` instance gets freed by caller `ossl_qlog_set_sink_bio()`